File: cracking_the_coding_interview/2_2_kth_to_last/kth_to_last.cpp
#include <iostream>
#include <vector>
#include <cassert>
#include <tuple>
#include <list>

// Finds the kth to the last element of the provided singly linked list.
//
// time complexity: O(N)
// space complexity: O(1)
std::list<int>::const_iterator kth_to_last(const std::list<int>& list, size_t k) {
    auto it1 = list.begin();
    auto it2 = list.begin();

    // Forward it1 k times.
    for (size_t i = 0; i < k; ++i) {
        if (it1 == list.end()) {
            return it1;
        }
        ++it1;
    }
    if (it1 == list.end()) {
        return it1;
    }

    // Now it1 is k positions in advance of it2. When it1 is exhausted
    // it2 is at the kth to the last position. However we must take care
    // to advance it2 not if it1 is already exhausted (at list.end()).
    while(++it1 != list.end()) {
        ++it2;
    }

    return it2;
}

// test implementation

std::ostream& operator<<(std::ostream& os, const std::list<int>& list) {
    os << "[";
    for (auto e: list) {
        os << e << ",";
    }
    os << "]";

    return os;
}

std::ostream& operator<<(std::ostream& os, const std::list<int>::const_iterator it) {
    os << *it;
    return os;
}

// Find the ith element of the provided list.
std::list<int>::const_iterator get_ith_node(const std::list<int>& l, int i) {
    if (i < 0) {
        return l.end();
    }

    auto it = l.begin();

    // Iterate through the list until the ith element or end of the list
    for (size_t index = 0; index < i && it != l.end(); ++index) {
        ++it;
    }

    // If i is larger than the list size, it will return list.end()
    return it;
}


#define ASSERT(expression)                               \
    if(!(expression))                                    \
    {                                                    \
        std::cerr << "Assertion failed: " << #expression \
                  << "\nFile: " << __FILE__ << "\n"      \
                  << "\nLine: " << __LINE__ << "\n";     \
        return false;                                    \
    }

#define ASSERT_EQUAL(a, b)                                \
    if((a) != (b))                                        \
    {                                                     \
        std::cerr << "Assertion failed: " << #a << " == " \
                  << #b << "\n" << #a << " = " << (a)     \
                  << ", " << #b << " = " << (b)           \
                  << "\nFile: " << __FILE__ << "\n"       \
                  << "\nLine: " << __LINE__ << "\n";      \
        return false;                                     \
    }


bool run_tests() {
    using vec = std::vector<int>;

    std::vector<std::tuple<vec, int>> test_cases;
    test_cases.emplace_back(vec{}, 2);
    test_cases.emplace_back(vec{1}, 0);
    test_cases.emplace_back(vec{1}, 1);
    test_cases.emplace_back(vec{1, 1}, 1);
    test_cases.emplace_back(vec{1, 2}, 0);
    test_cases.emplace_back(vec{1, 2}, 1);
    test_cases.emplace_back(vec{1, 2}, 2);
    test_cases.emplace_back(vec{1, 2, 3, 4, 1, 2, 3, 4}, 0);
    test_cases.emplace_back(vec{1, 2, 3, 4, 1, 2, 3, 4}, 1);
    test_cases.emplace_back(vec{1, 2, 3, 4, 1, 2, 3, 4}, 5);
    test_cases.emplace_back(vec{1, 2, 3, 4, 1, 2, 3, 4}, 100);

    for (auto& [input, k]: test_cases) {
        std::list<int> l(input.begin(), input.end());
        auto kth = kth_to_last(l, k);
        auto expected = get_ith_node(l, static_cast<int>(input.size()) - k - 1);
        if(expected == l.end()) {
            ASSERT_EQUAL(kth, l.end());
        } else {
            ASSERT_EQUAL(*kth, *expected);
        }
    }

    return true;
}


int main() {
    if(run_tests()) {
        std::cout << "C++ test succeeded\n";
        return 0;
    } else {
        std::cout << "Failure in C++ tests\n";
        return 1;
    }
}

File: cracking_the_coding_interview/2_2_kth_to_last/kth_to_last.py
import unittest

class ListNode:
    def __init__(self, value=None):
        self.value = value
        self.next = None

    def __str__(self):
        res = f"('{self.value}'"
        it = self.next
        while it:
            res += f"->'{it.value}'"
            it = it.next
        res += ")"

        return res


# Finds the kth to the last element of the provided singly linked list.
#
# time complexity: O(N)
# space complexity: O(1)
def kth_to_last(head: ListNode, k: int) -> ListNode:
    it1 = head
    it2 = head

    # Forward it1 k times.
    for i in range(k):
        if not it1:
            return None
        it1 = it1.next

    if not it1:
        return None

    # Now it1 is k positions in advance of it2. When it1 is exhausted
    # it2 is at the kth to the last position.
    while it1.next:
        it1 = it1.next
        it2 = it2.next

    return it2

# test implementation

# Create a singly linked list from given array.
def make_list(values) -> ListNode:
    if not values:
        return None

    head = ListNode(values[0])
    current_node = head

    for value in values[1:]:
        current_node.next = ListNode(value)
        current_node = current_node.next

    return head

# Check if two singly linked lists are equal.
def list_equal(node1: ListNode, node2: ListNode):
    if not node1 and not node2:
        return True
    elif node1 and not node2:
        return False
    elif node2 and not node1:
        return False

    it1 = node1
    it2 = node2

    while it1 and it2:
        if it1.value != it2.value:
            return False

        it1 = it1.next
        it2 = it2.next

    return it1 == it2

# Return the ith node from given list.
def get_ith_node(head: ListNode, ith: int) -> ListNode:
    if ith < 0:
        return None

    it = head
    for i in range(ith):
        if not it:
            return None
        it = it.next

    return it


class KthToLastTest(unittest.TestCase):
    def test_kth_to_last(self):

        test_cases = [
            ([], 2),
            ([1], 0),
            ([1], 1),
            ([1, 2], 0),
            ([1, 2], 1),
            ([1, 2], 2),
            ([1, 2, 3, 4, 1, 2, 3, 4], 0),
            ([1, 2, 3, 4, 1, 2, 3, 4], 1),
            ([1, 2, 3, 4, 1, 2, 3, 4], 5),
            ([1, 2, 3, 4, 1, 2, 3, 4], 100),
        ]
        for values, k in test_cases:
            l = make_list(values)
            kth = kth_to_last(l, k)
            expected = get_ith_node(l, len(values) - k - 1)

            self.assertEqual(kth, expected)



if __name__ == "__main__":
    unittest.main()
